### ISSUE TITLE
Please Fix ASAP!!!!! NSURL Session issue likely causing my entire app to crash

### ISSUE THREAD

**Issue by sdevo619** [particle-iot/particle-cloud-sdk-ios#15]
Can you pretty please, fix the URL Session and Crashing issues with the SDK. I don't mean to come off harsh here, but this SDK is not being maintained, and every year appears to be falling further behind the ball. 

Do you need help maintaining or would you like some? I need this repo and the Particle Setup one kept up to date, if I am going to be able to keep using them. If you do not plan to keep them up to date going forward, please just let me know, as this has gotten out of hand at this point, with how neglected this SDK is getting.

You can reference my conversation with Matt, if you would like. Again, sorry for the rather harsh words. Every year this seems to get worse though. If you need help or are too busy with other stuff at Particle, to keep this up to date regularly, please just let me know, and I will build my own, as this just getting to be more trouble than it is worth.

iOS Background Issues with your SDK( Post iOS 13+):

**The error I keep seeing, is: Access token is not found.

***Regarding this issue itself. The issue primarily occurs, while the device is locked, with my app in the background(best way to replicate, is on actual device, and run app without phone connected to the computer, as with the simulator running, you have unlimited background runtime vs. the 30 secs you have normally in the real world, which is when the issue is happening. For example, a notification will come in waking the app in the background, or if I use the today extension widget, to run a function in the main app, is when I am seeing this error message regarding the token. Again, this all was working flawlessly, prior to the iOS 13 rollout. Please help!!!
Thank you very much!!!
-Spencer

Additional Info for Dev Team:
Here is my best guess, as to the most likely cause of this issue currently, as well as more issues with your SDK. I believe it may be related to outdated code used to handle URL Sessions, however I have not dug through the issues deep enough to be certain, but again, actually keeping this updated regularly before it’s totally broken would have helped prevent this issue, as this doesn’t even begin to cover the Particle Setup Library issues either.

Main Issue Maybe??:
![image](https://user-images.githubusercontent.com/7008354/66865296-ca78a280-ef4b-11e9-8ca6-411d54c18f60.png)

Other Issues:
![image](https://user-images.githubusercontent.com/7008354/66865322-d82e2800-ef4b-11e9-99de-0ef931042520.png)

![image](https://user-images.githubusercontent.com/7008354/66865337-e4b28080-ef4b-11e9-8ee2-c60e96f74ddc.png)

**Comment by RaimundasSakalauskas** [comment]
Hello, thank you for reaching out. SDK is maintained and working as expected. The screenshots you posted above are just the warnings. I will look into background issue, repro project would help me get started and possibly look into that sooner.

**Comment by sdevo619** [comment]
Thank you for the fast response!!! I sincerely appreciate it. This is almost the exact same issue that happened with your SDK several years back, but before it was on launch or on foreground entry, now it is only impacting the device when locked. I was left hanging for months before it was fixed by your team finally, so please understand where I am coming from here.

From what I can gather right now, the issue is only present, when the device is locked, and a URL session is initiated, through your SDK ONLY, which has nothing to do with my app. All other URL Sessions initiated from within my own app are not responsible for this, and have been thoroughly tested for background issues, and are not presenting any problems. If I had to guess, it would be that Apple instituted some new background URL policies, that may have broken your code that has not been updated for changes in iOS 13 yet.

Also regarding those being warnings. They have been left in both the Particle SDK and Particle Setup Libraries for years now. So it's very difficult to believe what you are saying. I have been working with this SDK for several years now. Every single iOS release, the warning list, grows longer and longer. I see the growing list of them every time I compile too, which in itself is annoying to have to minimize constantly. Also, just because it is a warning, does not mean that it should be left for years like that. What if you did that with the check engine light on a car? The same principle applies here. They are warnings for good reason too. Please can you just fix the warnings, they've been left for way too long now. If I unlock the pod folder files and do it myself, I could be blamed for stability issues like this, which is why I have kept my hands out of it.
Thank you!!!!!!!!!!!!!!!

**Comment by RaimundasSakalauskas** [comment]
Do you have a demo project that could be used to test the issue? This being a rather edge case does not help the issue to get the attention it most likely deserves. Just want to let you know that I appreciate you making me (and Particle in general) aware of some of the issues we have with this SDK. Please also note that you are the first person to complain about these issues (and SDK works fine in Tinker, though we don't use background modes there) so having a project that would help me to reproduce this, would help me to get this issue bumped on the priority list. 

The issue will most likely be related to this as we use KeyChain to store session data: https://stackoverflow.com/questions/10536859/ios-keychain-not-retrieving-values-from-background

**Comment by sdevo619** [comment]
Thank you again!!! I really appreciate that, and your fast response. Now I already feel like we are heading in a better direction here. I told Matt, that if you'll take it, I will happily buy you a nice dinner for helping to fix this and all the warnings in both SDKs. Just let me know where you would like a gift card to. I am not a total prick, LOL, but I do need this working ASAP please!!!!

Unfortunately, due to the nature of what I am building, I cannot just send you a copy of the project. I can tell you, that the function that is being called every time that error is generated, is this one:

ParticleCloud.sharedInstance().getDevices { (particleDevices, error) in

Again, this also points to NSURLSession issues being the culprit here. I have no issue while my app is in the foreground, and performing a background refresh with the device connected to the simulator, works without any issues, but if I lock my device with it not running in the simulator, wait for a scheduled background refresh to run, which recalls the function above, iOS immediately kills my app. Again, this was all working beautifully on iOS 12 for me, so Apple may have changed something that is impacting your code related NSURLSession, is my best guess, and if you look at the warnings, some of the NSURL Related ones may be the cause or related to this error. I have this deployed through CocoaPods too, just FYI. This is also a product too, not just a regular device.

**Comment by sdevo619** [comment]
Also as far as the keys, or rather ID and Secret. I am the only dev currently working on this, I have the keys hard coded for testing purposes, to prevent having to troubleshoot key storage issues like this one here too, so that is one thing we can eliminate off the bat, at least on my end.

**Comment by RaimundasSakalauskas** [comment]
Thank you for your offer, but that would introduce a very bad practice where someone would have to buy me dinner every time they want something fixed :) Let's just use official process here. 

I understand that you cannot share your project source code and I don't want that responsibility either. What I am asking is if you could create a minimum sample project that would allow me to simply run it and experience the crash. This is a common practice by companies such as Unity to ask for sample project that would help reproduce the issue if it's not easily reproducible.

**Comment by sdevo619** [comment]
No worries about the meal. I just can't be waiting months on fix from you guys like the last time this happened this same error showed up, and I was just trying to grease the wheels here, and show you how much I appreciate your time and energy on this. Again, thank you very much, for the quick responses. I also believe, that the last time this error message began popping up, and presented itself the fix was also URLSession related.

I hope that now you understand how important this is to get it prioritized, as this issue, has made my application completely unreliable, to the point, to where I am considering abandoning your iOS SDK altogether, and building out my own URLSession code at this point. 

Just a few months ago, I spent multiple days and countless hours troubleshooting issues with Particle Borons going back and forth with your team, constantly claiming it was something on my end, when it never was, and bad firmware updates released by your team, and every single time here I am running into issues, it seems to be with issues with things on your end, that I have zero control over, making this whole process feel like insanity, both as a customer and even more so as product creator, as I never want customers to have to go through this much to get a single thing fixed.

I understand what you are asking for, and if that is the only way I can get this fixed, I will. Please understand though, that it is going to take at the very least, a few hours of my time, to recreate this in a new project, as it is a product, add all the bridge header code and settings, and I will also have build it with background refresh and then test everything, vs. the rather small amount of time it will take, to fix the warnings throughout both Libraries, or the very least the Particle SDK one, which I currently believe to be at fault anyways, but both are way overdue to be fully updated, by years now. You can even send a branch to test if you are not comfortable pushing it out to everyone without more time to test internally. If it doesn't fix the issue, I will gladly eat my words here.

If you are willing to work on that stuff, and begin updating it, I will gladly take the time to create an example project for you to test. I am happy to help you with any testing, as long as I feel like I am not the only one doing all the work here. I hope this makes sense. Thank you very much again!!!

**Comment by RaimundasSakalauskas** [comment]
The things you list as "problems" are just compiler warnings. Those are really low priority and I know that those do not cause any crashes and won't be fixed immediately. I hope I can get them fixed at some point, but I know that won't happen anytime sooner than December (and I can't make any promises for that time either; just trying to set short term expectations).

Thank you for your effort in trying to help us. Please also understand that the issue you are seeing with the crashes affects very little of our customers (I wouldn't be surprised if you are the only one who faces the background crash issue), therefore as much as I would like to work on this today - I simply can't as there are more pressing issues that need to be addressed. If you could help me by creating a sample project that would trigger the background crash you described, it would help me with getting approval for time needed to work on this.

**Comment by RaimundasSakalauskas** [comment]
https://github.com/particle-iot/particle-sdk-ios/tree/feature/fix-background-session-crash - 
here's a quick attempt at fixing a crash based on the information you submitted above (I didn't have time to work on an example project that would help me reproduce the issue). I have added a change to development branch. Can you please try using this branch and let me know if this fixes your issue? You might need to log out and log in to the app for this to take affect. 

To use development branch with cocoapods you should update your podfile:
`pod 'Particle-SDK', :git => 'https://github.com/particle-iot/particle-sdk-ios.git', :branch => 'feature/fix-background-session-crash'`

**Comment by sdevo619** [comment]
I'm very confused, and bit frustrated as to how you can justify leaving it until December as the iOS Dev there, but I guess that is how you guys take your customers feedback, which really hurts without how much effort I have put in here already. 

Yes there are much more pressing things to do with this SDK, and I could give you list that stretches a mile long of what it needs, as it needs a ton, but you guys don't seem to care about that stuff, as not a single thing I have ever mentioned, such as bringing two factor auth, which should be at the top of the list, has been added yet, and keep in mind, that not including that by now, with how many times I alone have asked for it, essentially makes just about any potential IoT Product, from any  creator using this SDK, not just my own stuff, dead on arrival. 

I know that it can be added separately, but there is a much bigger issue here. Your own accounts have it, and have had it for what seems like quite some time now, but the products creators, who have the most to lose if hacked, at least who use the iOS SDK, are SOL.

Here is the kicker, that as far as I am aware, I still cannot turn it on, on my product creator account, to even protect that, without breaking the iOS SDK completely. That is not just an oversight, and extremely dangerous, and it comes off that you don't give a crap about the products creators security, even if not intended that way. 

It is also a huge liability on your company's part too, so it might be worth pushing it up the priority list. Please correct me if I am wrong here, but the last time I enabled it, it broke the SDK completely, and the only fix was turning it back off.

You are not making the best case for continuing to use the SDK at this point, or even to continue working with your company, which is really sad for me, as I see so much potential, and even wanted to work with you guys, at Particle. 

The amount of effort you require on your customer's part, here being a prime example, to even put any effort in solving these simple issues or improving the SDK at all, is exactly why more people just are not willing to take this much time out of their lives to tell you about these problems, because they simply don't have it, and will just move on to a different SDK, micro controller, or company, and you guys lose in the end, not me. 

Back to the issue at had though. If nobody else is reporting it, it's because either not enough people are using the SDK, or they aren't building products on the scale I am, or with the extensive feature list mine has, and the more of those I sell, the more units you sell, so I don't get why you keep fighting me on this stuff, as we can both win together here. I will build my own Libraries out in Swift if I have to. Please don't think for a second I am not capable.

If you don't have the time, can you please ask you boss to hire another dev to help you out, as these are not big things I am asking for, and would have taken far less time, than having to respond to all these messages.

In the time since my last response, I dug into the "warning" messages, that you say could not possibly be causing this, and before I waste multiple more hours of my own time, please examine these, as they both more the explain what are causing these issues. 

If you won't take the time to fix these simple things, I will be left with no choice, but to put you guys on blast in the forums again. I will reiterate, that I do not want to do this. If you fix these two things and it doesn't fix the problem for me, I will gladly send you a demo project before saying anything in the forums, and call you the King of iOS Dev.

I have asked nicely enough times though now, and I am getting really tired of having to explain this to someone, who I know should know better than this. If you just look at these two specific warnings, they are glaringly obvious as to what has broken your SDK and causing these issues for me, and are without question contributing to my problems and broken app. 

I do not like having to resort to these kinds of tactics, and following this experience, I will more than likely, either just be abandoning your controllers as soon as possible, or will be writing my own SDK in Swift, as you clearly don't give a crap about your product creators, or at least not enough of one, to take any of their feedback seriously. 

I do not want to have to be the pain in the ass customer or threaten to go over your head here, but your lack of effort, is leaving me very little choice in the matter. I will ask again, as nicely as I can, please fix this stuff NOW!!!!

<img width="1641" alt="Screen Shot 2019-10-23 at 12 16 03 PM" src="https://user-images.githubusercontent.com/7008354/67426817-bb1ed800-f58f-11e9-8b41-320814abbbc4.png">
<img width="1649" alt="Screen Shot 2019-10-23 at 12 13 32 PM" src="https://user-images.githubusercontent.com/7008354/67426819-bb1ed800-f58f-11e9-9f9a-edfa5ff36a1e.png">

**Comment by sdevo619** [comment]
Thank you very much for making a new branch, I had not reloaded the page, and just saw your last message. I am not able to install the new branch though. I get an error message every time I try to, so no luck so far. Just to confirm, I did comment out the other Particle SDK reference in the pod file.
Error message:

[!] There are multiple dependencies with different sources for `Particle-SDK` in `Podfile`:

- Particle-SDK (from `https://github.com/particle-iot/particle-sdk-ios.git`, branch `feature/fix-background-session-crash`)
- Particle-SDK

**Comment by busticated** [comment]
hi @sdevo619 popping in here to say: we very much hear you and appreciate your feedback. i know how frustrating it can be to have blocking bugs and i can assure you we take these issues very seriously.

i need to ask you to cool your tone a bit here though - take a step back and try to remember we're all humans doing our best :)

in the meantime, know that we're very much aware of some of the rougher spots of our offering and are working every day to iron those out. obviously i can't commit to specific dates but know that great iOS support is _critical_ to our business. please do share your feedback and feature requests on our forums ([link](https://community.particle.io/c/mobile)) as that's where they'll gain the most visibility from the broader team and other users.

as for the issue w/ warnings and background crashes, let's continue working on it here.

thanks :pray:

**Comment by sdevo619** [comment]
If I actually felt like this stuff was being taken seriously, I would gladly spend more time telling you how to improve. His last response prior to mine that you are referring to, that you think was too harsh, made it seem like this was not a priority and he would get to whenever, unless I could build him a test project, after I have spent the last month and way too much time now, telling both him and your team exactly where to look. 

The reason more people don't report these kinds of issues to you, is because of what they have to go through to get them any attention. None of what you said makes me feel, like a single piece of feedback I have given your team through countless emails now, has gone anywhere but the garbage, so why should I waste anymore on the forums, reporting it there, as I am still getting a ton of resistance to even fix two lines of code??

If you would consider, just assigning an actual dev to oversee cases like this, who deals specifically with product creators, someone who is not just support engineer, someone with in depth knowledge of how to actually fix this kind of stuff, or who in a worst case scenario, can email the actual person directly, who is fixing it, and tell them to fix this in the next release and push up the next release if need be, as one of our creators apps, is crashing constantly because of it, and notifications are not being delivered as result. At the very least, they can tell the iOS dev to spend some time on this ASAP and that it needs to be made a priority. 

I would not have had to now explain this same issue to your main support staff, and again here, and still have to spend extra time convincing both people, in both places, and now you as well, to the importance of getting this fixed ASAP, and also would not have had to explain the backstory of all the other mistakes I have had to field from your company, just in the last year alone, making everyone get way more of their time back and keeping the stress levels way down. On top of that, you would sell more Particle Devices, as both our products would not only give the customer a better experience, their perceived reliability, would go up exponentially, again leading to more sales. Win win.

I do not ever expect you to be perfect. Being so unwilling to admit to a mistake, that you are completely apathetic to the person taking the time to ask for help, as well as help your team, with a serious reliability issue, is disappointing to say the least. It is very discouraging as product creator and someone who loves particle enough to dedicate this much of his life, to building an insanely cool product with it. At least that was originally how the response prior to the one with the new repo in it, came off, even if not intended that way.

Seeing your message, with a new branch was an enormous surprise, as it seemed to have been dismissed in my prior message, even if it is not working because of that error, I really needed to see, that some effort was being made in a somewhat timely fashion, and was having trouble seeing that, until I saw that message. I wish I had seen before sending my last message, but GItHub didn't show it until after I sent that. I am sorry if what I said came off too harsh, and thank your for taking the time to build a new branch. I can only base my responses, on what my continued experience with your company has been until now, and your prior response did not make me feel like this was going to get any attention. I am always happy to be wrong about that. Again, thank you!!!!

Your most recent firmware rollout issues, just a fews month back, completely broke all my Borons for over a month, completely killed my Electron, which still doesn't work as a result of it, and after countless hours going back and forth with your engineers, who all were claiming it was on my end, it wasn't until I put your team on blast in the forum, and countless others began to chime in, that you actually stepped up and took responsibility for a single thing here. 

I absolutely hate doing things like that, and problems should not have to get to that point to get attention, but if a company won't listen to it's customers, especially ones who could easily lead to the sales of thousands of units, the ones your company has positioned themselves as advocates for and as teammates with, especially when they are saying there is an issue over and over again, and you keep telling them/me, that it has to be on my end, without looking into the issue thoroughly yet, or at all, you leave the customer, with very few good options. In the case of product creators, someone or even a group of people, who are betting their life and well being and others too, on you delivering, you are spitting in their faces.

Even after me doing that, your team left the Borons without working firmware for over a month making them completely useless for that time, so your track record was not helping with my confidence that this issue was going to get the attention it needed, and why my reactions are not getting better as time goes on here. I am still playing catch up from that huge mess, it was such a problem for the company I am building. I believe that the previous year, almost the identical issues was left unfixed for multiple months, before it was finally fixed. It may have been the prior year, but those are just a couple of my experiences with your team. I have lots more examples of bad ones unfortunately. I did not arrive at this level of frustration over one single mistake, it was quite a few of them. 

I would really like at this point, for this to be the start to some better ones though, if you can please fix whatever is preventing Cocoa Pods from replacing the current SDK with branched one you want me to test, or tell me what I need to do to fix it, if it is an issue on my end.  Hopefully we can move forward in a better direction. I already tried to clean the project which did not help. I will happily test the branch and give you constructive feedback on the changes, as I know it is now being worked on. I did not ever expect you to drop everything for me and was and am willing to pay, for an hour of his time if I have to, to fix this, but I do need it fixed, not just promises that we'll look into it.
Again, Thank you, I'm sorry if I came off too harsh, but hopefully you can now better see, how I got here. All it takes is one person to change that though. Hopefully it's you!!!!

**Comment by RaimundasSakalauskas** [comment]
> Thank you very much for making a new branch, I had not reloaded the page, and just saw your last message. I am not able to install the new branch though. I get an error message every time I try to, so no luck so far. Just to confirm, I did comment out the other Particle SDK reference in the pod file.
> Error message:
> 
> [!] There are multiple dependencies with different sources for `Particle-SDK` in `Podfile`:
> 
> * Particle-SDK (from `https://github.com/particle-iot/particle-sdk-ios.git`, branch `feature/fix-background-session-crash`)
> * Particle-SDK

Make sure your podfile has only 1 entry for Particle-SDK and update it using `pod update Particle-SDK` command. Looking at the error seems that you included Particle-SDK there twice. If this doesn't help, please post your podfile as the issue above has nothing to do with the dependency itself, but rather your project setup.

**Comment by sdevo619** [comment]
Ok guys. So after my level of frustration with current state of the SDK reaching a boiling point. And your branch not working for me, also there was not a duplicate entry in my POD file. I decided to take different approach. I had previously contacted the dev who built Particle Swift, but I never received a response, and his Libraries had not been updated in several years and no longer functioned correctly with iOS 13, so I did what I said I was going to do. 

In less than 3 days, using Particle Swift for a reference, where your documentation was lacking, I built a fully functioning Library, entirely in Swift, yes you read right, entirely in Swift, that handles tokenization, logins, and can handle reads and writes, and I will be adding subscriptions this next week. 

I am not here to gloat, or to talk crap, I am writing this now, to thank you, as I am not sure that I would have had the determination to do it, if your SDK had not totally broken for me. While this may not be the solution you were shooting for, it's far better than what I would have received in your update, as I now have control over all the pieces that were driving me nuts about your SDK. Also, no more Objective C Code or bridging headers required!!!!!!!!!!

While I would still highly recommend adding/updating everything I asked for and have asked for, as it will greatly improve your company's abilities to launch products and set up those company's launching them, for success, at this point, all I need from your team now, is to push out reliable firmware updates going forward, and to keep the backend healthy, and you won't hear any complaints on my end.

Again, I am sorry things got heated, but if you knew what I had riding on this, and what I had gone through with your company to get here, you would have not made me work so hard, to get some support. As I said before, the large majority of people will just call you product garbage and say your company doesn't care about its customers and just move on, which is why you don't hear more of this stuff. 

I hope you see this as an opportunity to do better the next time, and not treat one of your greatest assets, like they don't know what they are talking about, until they lose their patience, as that has been my experience until now with your team. If you do ever decide to rework the SDK and update it, I would be happy help with feedback and whatever else I can. We can win much better together, then we can on our own.

Again, thank you!!!!!!!!! Take care!!!!!!
